version-published/snap_info_utility.py

```diff
@@ -20,3 +22,39 @@ def get_snap_info_from_store(snap_name: str) -> dict:
         )
 
     return response.json()
+
+
+def get_history_since(tag: str, repo_path: str):
+    return check_output(
+        [
+            "git",
+            "log",
+            "--pretty=format:%H",
+            "--no-patch",
+            f"{tag}~1..origin/main",
+        ],
+        text=True,
+        cwd=repo_path,
+    ).splitlines()
+
+
+def get_offset_from_version(version: str) -> int:
+    if "dev" not in version:
+        return 0
+    return int(version.rsplit("dev", 1)[1])
+
+
+def get_revision_at_offset(version: str, repo_path: str):
+    tag = get_latest_tag(repo_path)
+    history = get_history_since(tag, repo_path)
+    offset = get_offset_from_version(version)
+    # history is now -> tag(inclusive)
+    # raw tag (no dev) is tag so -0 -1 = -1 (latest in history)
+    # dev-1 is tag+1 so -1 -1 = -2 semi lastest in history
+    return history[-offset - 1]
```

Comment:
This smells unsafe.

history IIUC is a list of strings.
I understand that offset is positive, this is why you use a unary `-`. But in the end there's a lot of opportunities here to get a negative index. 

I'm guessing the correctness can be inferred from the relationships between functions, but this is true only "right now". Boundary checks here would make it easy on the next person. Or assertions.

Reply:
Also the comment here suggests that the index should be negative ? :confused:

Reply:
offset is positive (as it is the distance from the tag). Given that we have an history from now to the tag, we have to use a negative index to "add" to the tag as:
- history[-1] is the tag
- history[-2] is the commit after the tag
- hitory[0] is the current HEAD

I will add the boundary checking now